### PR TITLE
Add tuco library with its tests

### PR DESCRIPTION
**Pull request by jaysonsantos**

### CODE REVIEW

**Comment by jaysonsantos** [comment]
Travis working but the job docs will be broken until we release the frist version https://travis-ci.org/eatfirst/python-tuco/jobs/325114115

File: .travis.yml [hunk 1]
```diff
@@ -19,10 +18,14 @@ matrix:
     - python: '3.6'
       env:
         - TOXENV=py36,report,coveralls
-    - python: 'pypy-5.4'
+    - python: '3.6'
```
Comment (andreoliwa) [inline .travis.yml]: Why 3.6 twice (lines 18 and 21)?

File: README.rst [hunk 2]
```diff
@@ -78,8 +78,7 @@ https://python-tuco.readthedocs.io/
 
 Development
 ===========
-
-To run the all tests run::
+Make sure you have a running redis instance and to run the all tests run::
 
     tox
```
[the remark this answers is not in the document]
Comment (andreoliwa) [reply README.rst]: I installed Python 3.4.7, [after setting some compilation flags on MacOS](https://github.com/pyenv/pyenv/wiki/Common-build-problems):

```
CFLAGS="-I$(brew --prefix openssl)/include -I$(xcrun --show-sdk-path)/usr/include" \
LDFLAGS="-L$(brew --prefix openssl)/lib" \
pyenv install -v 3.4.7
```

But still no success with tox:

```
ERROR:   py34: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.4: b"pyenv: python3.4: command not found\\n\\nThe `python3.4\' command exists in these Python versions:\\n  3.4.7\\n\\n"',)
ERROR:   py35: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.5: b"pyenv: python3.5: command not found\\n\\nThe `python3.5\' command exists in these Python versions:\\n  3.5.4\\n\\n"',)
ERROR:   py36: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
ERROR:   report: commands failed
ERROR:   docs: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
```

Full log:
```
$ tox
GLOB sdist-make: /Users/wagneraugusto/eatfirst/python-tuco/setup.py
clean installed: coverage==4.4.2
clean runtests: PYTHONHASHSEED='2128799634'
clean runtests: commands[0] | coverage erase
check installed: bleach==2.1.2,check-manifest==0.36,docutils==0.14,flake8==3.5.0,flake8-quotes==0.13.0,html5lib==1.0.1,isort==4.2.15,mccabe==0.6.1,mypy==0.560,psutil==5.4.3,pycodestyle==2.3.1,pyflakes==1.6.0,Pygments==2.2.0,readme-renderer==17.2,six==1.11.0,typed-ast==1.1.0,webencodings==0.5.1
check runtests: PYTHONHASHSEED='2128799634'
check runtests: commands[0] | python setup.py check --strict --metadata --restructuredtext
running check
check runtests: commands[1] | check-manifest /Users/wagneraugusto/eatfirst/python-tuco
lists of files in version control and sdist match
check runtests: commands[2] | flake8 src tests setup.py
check runtests: commands[3] | isort --verbose --check-only --diff --recursive src tests setup.py

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.2.15

\########################################################################/

SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/properties.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/__init__.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/exceptions.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/graph_builder.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/base.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/meta.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/decorators.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/locks/memory.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/locks/__init__.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/locks/redis.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/src/tuco/locks/base.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/tests/test_tuco.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/tests/example_fsm.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/tests/__init__.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/tests/test_graph_creation.py Everything Looks Good!
SUCCESS: /Users/wagneraugusto/eatfirst/python-tuco/setup.py Everything Looks Good!
check runtests: commands[4] | mypy src tests setup.py
py34 create: /Users/wagneraugusto/eatfirst/python-tuco/.tox/py34
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.4: b"pyenv: python3.4: command not found\\n\\nThe `python3.4\' command exists in these Python versions:\\n  3.4.7\\n\\n"',)
py35 create: /Users/wagneraugusto/eatfirst/python-tuco/.tox/py35
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.5: b"pyenv: python3.5: command not found\\n\\nThe `python3.5\' command exists in these Python versions:\\n  3.5.4\\n\\n"',)
py36 create: /Users/wagneraugusto/eatfirst/python-tuco/.tox/py36
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
report installed: coverage==4.4.2
report runtests: PYTHONHASHSEED='2128799634'
report runtests: commands[0] | coverage combine --append
No data to combine
ERROR: InvocationError: '/Users/wagneraugusto/eatfirst/python-tuco/.tox/report/bin/coverage combine --append'
docs create: /Users/wagneraugusto/eatfirst/python-tuco/.tox/docs
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
_________________________________________________________________________________________________ summary __________________________________________________________________________________________________
  clean: commands succeeded
  check: commands succeeded
ERROR:   py34: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.4: b"pyenv: python3.4: command not found\\n\\nThe `python3.4\' command exists in these Python versions:\\n  3.4.7\\n\\n"',)
ERROR:   py35: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.5: b"pyenv: python3.5: command not found\\n\\nThe `python3.5\' command exists in these Python versions:\\n  3.5.4\\n\\n"',)
ERROR:   py36: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
ERROR:   report: commands failed
ERROR:   docs: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' command exists in these Python versions:\\n  3.6.2\\n  3.6.3\\n\\n"',)
```

I will investigate it later.

File: .travis.yml (hunk 1 above)
Comment (jaysonsantos) [reply .travis.yml]: This is a matrix, so it is python 3.6 and env TOXENV=docs so it has a dedicated docs tasks

File: README.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaysonsantos) [reply README.rst]: as you are using pyenv it changes the paths so you have to run this before `pyenv local 3.4.7 3.6.2 3.5.4`

[the remark this answers is not in the document]
Comment (jaysonsantos) [reply README.rst]: At the end of the message it shows the pyenv output

> ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError('Failed to get version_info for python3.6: b"pyenv: python3.6: command not found\\n\\nThe `python3.6\' **command exists in these Python versions:\\n  3.6.2\\n  3.6.3**\\n\\n"',)

File: .travis.yml (hunk 1 above)
Comment (andreoliwa) [reply .travis.yml]: Hmm... so using `- TOXENV=py36,report,coveralls,docs` generates a different output?
I'm trying to understand how it works.

File: README.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreoliwa) [reply README.rst]: > so you have to run this before pyenv local 3.4.7 3.6.2 3.5.4

I don't know how to do that, never mind.
It's working, so all good.
I will read more about tox in the future.
And people that want to fork Tuco will probably know how to use tox.

File: .travis.yml (hunk 1 above)
Comment (jaysonsantos) [reply .travis.yml]: No, it will create one TOXENV=py32,report,coveralls and one with TOXENV=docs both running with travis python3.6 support

**Pull request merged**